### PR TITLE
Merge master into new-katstore-integration feature

### PR DESCRIPTION
**Pull request by tockards**
This PR merges master into the katstore feature. 

This branch has to be fully tested first I am making the PR for review purposes. 

The major differences are the 
 - python 3 compatibility 
-  change in the publishing of sensor data(no longer retrieves sensor data from redis)
- change references from ts to time

### CODE REVIEW

File: katportalclient/test/test_client.py [hunk 1]
```diff
@@ -921,36 +968,118 @@ def test_sensor_value_one_result(self):
...
+                                                               status='nominal')}
+        res = yield self._portal_client.sensor_values(["some_other_sample_0",
+                                                       "some_other_sample_1"])
+        assert res == expected_result, res
+
     @gen_test
     def test_sensor_history_single_sensor_without_value_time(self):
-        """Test time ordered data without value_time for single sensor request."""
+        """Test that time ordered data with value_time is received for a
+        single sensor request.
```
Comment (ajoubertza) [inline katportalclient/test/test_client.py]: The old docstring was correct.  This test for "..._without_value_time"

File: katportalclient/test/test_client.py [hunk 2]
```diff
@@ -921,36 +968,118 @@ def test_sensor_value_one_result(self):
...
+        """
         history_base_url = self._portal_client.sitemap[
             'historic_sensor_values']
         sensor_name = 'anc_mean_wind_speed'
         publish_messages = [sensor_history_pub_messages_json['init']]
         publish_messages.extend(sensor_history_pub_messages_json[sensor_name])
```
Comment (ajoubertza) [inline katportalclient/test/test_client.py]: All the `publish_messages` parts can be removed (this test and the ones that follow).  They are left over from old katstore tests.
The `sensor_history_pub_messages_json` variable must be deleted too.

File: katportalclient/test/test_client.py [hunk 3]
```diff
@@ -1042,13 +1171,34 @@ def test_sensor_history_single_sensor_invalid_times(self):
...
+        publish_messages = [sensor_history_pub_messages_json['init']]
+        publish_messages.extend(sensor_history_pub_messages_json[sensor_name])
+
+        self.mock_http_async_client().fetch.side_effect = self.mock_async_fetcher(
+            valid_response='{"result":"success"}',
+            invalid_response='error',
+            starts_with=history_base_url,
+            contains=sensor_name,
+            publish_raw_messages=publish_messages,
+            client_states=self._portal_client._sensor_history_states)
```
Comment (ajoubertza) [inline katportalclient/test/test_client.py]: The `publish_raw_messages` and `client_states` must be removed - not used for new katstore.

Also remove `self._sensor_history_states = {}` from `client.py` - it is unused.

File: katportalclient/test/test_client.py [hunk 4]
```diff
@@ -1062,9 +1212,9 @@ def test_sensor_history_multiple_sensors_valid_times(self):
         #  - 1st call gives sensor list
         #  - 2nd call provides the sample history for sensor 0
         #  - 3rd call provides the sample history for sensor 1
-        self.mock_http_async_client().fetch.side_effect = mock_async_fetchers(
+        self.mock_http_async_client().fetch.side_effect = self.mock_async_fetchers(
             valid_responses=[
-                '{"data" : [%s, %s]}' % (sensor_json[sensor_names[0]],
+                '{"data" : [{}, {}]}'.format (sensor_json[sensor_names[0]],
                                          sensor_json[sensor_names[1]]),
```
Comment (ajoubertza) [inline katportalclient/test/test_client.py]: Indentation is needs to be fixed.

File: katportalclient/test/test_client.py [hunk 5]
```diff
@@ -1062,9 +1212,9 @@ def test_sensor_history_multiple_sensors_valid_times(self):
         #  - 1st call gives sensor list
         #  - 2nd call provides the sample history for sensor 0
         #  - 3rd call provides the sample history for sensor 1
-        self.mock_http_async_client().fetch.side_effect = mock_async_fetchers(
+        self.mock_http_async_client().fetch.side_effect = self.mock_async_fetchers(
             valid_responses=[
-                '{"data" : [%s, %s]}' % (sensor_json[sensor_names[0]],
+                '{"data" : [{}, {}]}'.format (sensor_json[sensor_names[0]],
```
Comment (ajoubertza) [inline katportalclient/test/test_client.py]: Shouldn't be whitespace between `format` and `(`.

File: katportalclient/test/test_client.py [hunk 6]
```diff
@@ -1565,44 +1722,86 @@ def mock_async_fetchers(valid_responses, invalid_responses, starts_withs=None,
     assert(len(starts_withs) == num_calls)
```
Comment (ajoubertza) [inline katportalclient/test/test_client.py]: `mock_async_fetchers` must be changed to a class method - see code in master branch.

File: katportalclient/test/test_client.py [hunk 7]
```diff
@@ -1565,44 +1722,86 @@ def mock_async_fetchers(valid_responses, invalid_responses, starts_withs=None,
...
                     for v, i, s, e, c in zip(
                         valid_responses, invalid_responses,
                         starts_withs, ends_withs, containses)]
     # flip order so that poping effectively goes from first to last input
     mock_fetches.reverse()
 
-    def mock_fetch(url):
-        single_fetch = mock_fetches.pop()
-        return single_fetch(url)
+        def mock_fetch(url):
```
Comment (ajoubertza) [inline katportalclient/test/test_client.py]: Indentation incorrect - see code in master branch.

File: katportalclient/test/test_client.py [hunk 8]
```diff
@@ -1565,44 +1722,86 @@ def mock_async_fetchers(valid_responses, invalid_responses, starts_withs=None,
...
+                single_fetch = mock_fetches[-1]
+            else:
+                single_fetch = mock_fetches.pop()
+            return single_fetch(url)
 
-    return mock_fetch
+        return mock_fetch
 
 
 def mock_async_fetcher(valid_response, invalid_response, starts_with=None,
```
Comment (ajoubertza) [inline katportalclient/test/test_client.py]: `mock_async_fetcher` must be changed to a class method - see code in master branch.

File: katportalclient/test/test_client.py [hunk 9]
```diff
@@ -1565,44 +1722,86 @@ def mock_async_fetchers(valid_responses, invalid_responses, starts_withs=None,
...
+                            raw_message = raw_message.replace(
+                                'test_namespace', namespace)
+                        test_websocket.write_message(raw_message)
+
+            body_buffer = buffer_bytes_io(response)
+            result = HTTPResponse(HTTPRequest(url), 200, buffer=body_buffer)
+            future = concurrent.Future()
+            future.set_result(result)
+            return future
         result = HTTPResponse(HTTPRequest(url), 200, buffer=body_buffer)
```
Comment (ajoubertza) [inline katportalclient/test/test_client.py]: These lines shouldn't be here any more - see code in master branch.

File: katportalclient/test/test_client.py [hunk 10]
```diff
@@ -974,7 +974,7 @@ def test_sensor_value_one_result(self):
         assert res == expected_result
 
         self.mock_http_async_client().fetch.side_effect = self.mock_async_fetcher(mon_response)
-        expected_result = SensorSampleValueTs(sample_time=222.222, value_time=111.111,
+        expected_result = SensorSampleValueTime(sample_time=222.222, value_time=111.111,
                                               value=43680.0, status=u'nominal')
```
Comment (ajoubertza) [inline katportalclient/test/test_client.py]: Indentation. PEP8.

File: katportalclient/test/test_client.py [hunk 11]
```diff
@@ -1062,14 +1062,12 @@ def test_sensor_values_multiple_filters_multiple_matches(self):
...
     @gen_test
     def test_sensor_history_single_sensor_without_value_time(self):
-        """Test that time ordered data with value_time is received for a
+        """Test that time ordered data without value_time is received for a
         single sensor request.
         """
         history_base_url = self._portal_client.sitemap[
             'historic_sensor_values']
         sensor_name = 'anc_mean_wind_speed'
-        publish_messages = [sensor_history_pub_messages_json['init']]
```
Comment (ajoubertza) [inline katportalclient/test/test_client.py]: The definition of `sensor_history_pub_messages_json`  on line 222 still needs to be removed.

File: katportalclient/test/test_client.py [hunk 12]
```diff
@@ -1741,9 +1705,9 @@ def mock_fetch(url):
         return mock_fetch
 
 
-def mock_async_fetcher(valid_response, invalid_response, starts_with=None,
-                       ends_with=None, contains=None):
-    """Returns a mock HTTP async fetch function, depending on the conditions."""
+    def mock_async_fetcher(valid_response, invalid_response, starts_with=None,
```
Comment (ajoubertza) [inline katportalclient/test/test_client.py]: Needs `self` as first parameter.

**Comment by tockards** [comment]
@ajoubertza thanks for the review. I am closing the PR and will put it through once I test it.

**Comment by ajoubertza** [comment]
@tockards - Good to test, but no need to close the PR.  I don't want to have to review the same thing again, so I don't want a new PR!

**Comment by tockards** [comment]
all tests are passing

File: katportalclient/test/test_client.py [hunk 13]
```diff
@@ -1046,7 +1045,7 @@ def test_sensor_history_single_sensor_without_value_time(self):
         # ensure time order is increasing
         time_sec = 0
         for sample in samples:
-            self.assertGreater(sample.sample_time, time_sec)
+            self.assertGreater(float(sample.sample_time), float(time_sec))
```
Comment (ajoubertza) [inline katportalclient/test/test_client.py]: As discussed - this cast is unexpected.  `sample_time` should already be a float.  Check implementation and test input.

File: katportalclient/test/test_client.py [hunk 14]
```diff
@@ -1167,15 +1166,15 @@ def test_sensor_history_multiple_sensors_valid_times(self):
         # expect keys to match the 2 sensor names
         self.assertIn(sensor_names[0], list(histories.keys()))
         self.assertIn(sensor_names[1], list(histories.keys()))
-        # expect 3 samples for 1st, and 4 samples for 2nd
-        self.assertTrue(len(histories[sensor_names[0]]) == 3)
-        self.assertTrue(len(histories[sensor_names[1]]) == 4)
+        # expect 4 samples for 1st, and 3 samples for 2nd
```
Comment (ajoubertza) [inline katportalclient/test/test_client.py]: This should stay as it was.  Results should be in sorted sensor name order.

**Pull request merged**